### PR TITLE
Auto assign correct device serial no to matching camera name

### PR DESCRIPTION
**Pull request by sumandeepb**
Added code for auto detecting and mapping correct realsense sensor based on installation orientation.

[removed JIRA link]

resolves https://www.wrike.com/open.htm?id=422274965

Link to Design Document: https://docs.google.com/document/d/1Dg6gszNfD36r9-z7hvgF-JKsA93_GhdvBopIPmeXlRo/

### CODE REVIEW

**Comment by sumandeepb** [comment]
@gauthamm I just realized that some of the code might not conform to the coding standards specified in the document https://docs.google.com/document/d/1m6B8DLkw4JVb6Kf8Ku4O0uCbc7VjnhcIxjLcvKkQFcU/

I will make the changes and push a new commit into this PR

**Comment by sumandeepb** [comment]
Updated the code as per review comments. Now the accel orientation need not be manually specified. It is auto computed from the specified tf2 realsense link for both top and bottom sensor.

**Comment by sumandeepb** [comment]
resolves https://www.wrike.com/open.htm?id=422274979

@gauthamm @ChristofDubs I have updated the code as per requested changes. Kindly review.

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 1]
```diff
@@ -11,12 +11,25 @@
...
+{
+    if (abs(a.x - b.x) <= error &&
+        abs(a.y - b.y) <= error &&
+        abs(a.z - b.z) <= error)
+    {
+        return true;
+    }
+
+	return false;
+}
```
Comment (ChristofDubs) [inline realsense2_camera/src/realsense_node_factory.cpp:25]: could shorten this to
`return abs(a.x - b.x) <= error && abs(a.y - b.y) <= error && abs(a.z - b.z) <= error;`

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 2]
```diff
@@ -57,12 +74,110 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+				try {
+					geometry_msgs::TransformStamped transformStamped = 
+									_tf_buffer.lookupTransform(_tf_link_name, "base_link", ros::Time(0));
+
+					tf2::Quaternion q_orig, q_rot;
+					// universal gravity vector
+					q_orig.setValue(0.0, -9.81, 0.0);
+					tf2::convert(transformStamped.transform.rotation , q_rot);
+					// Calculate the new orientation
+					q_orig = q_rot.inverse() * q_orig * q_rot;
```
Comment (ChristofDubs) [inline realsense2_camera/src/realsense_node_factory.cpp]: I would prefer if this was cleaned, such that it is more readable and the following questions resolve themselves:
- what is setValue with 3 arguments for a quaternion supposed to be?
- why is the gravity in the y entry?
- where did you get the q = Q^-1 * q * Q formula from?

The way I would clean this up would be like this (not sure if this will compile exactly like this):
```
tf2::Matrix3x3 rot_link_base;
tf2::convert(transformStamped.transform.rotation , rot_link_base);
const tf2::Vector3 base_g(0,0,9.81);
const tf2::Vector3 link_g = rot_link_base * base_g;
```

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 3]
```diff
@@ -57,12 +74,110 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+							no_attempts ++;
+							ROS_DEBUG_STREAM("Pipe failed to create on Device serial no " << sn << "device busy. Waiting..." << no_attempts);
+							ros::Duration(2).sleep();
+							continue;
+						}
+					}
+
+					if (!success)
+					{
+						// all max attempts exhausted for the current device
```
Comment (ChristofDubs) [inline realsense2_camera/src/realsense_node_factory.cpp:191]: consider to publish a warning in this case as this is not expected to happen, right?

File: realsense2_camera/include/realsense_node_factory.h [hunk 4]
```diff
@@ -66,10 +67,15 @@ namespace realsense2_camera
         virtual void onInit() override;
         void tryGetLogSeverity(rs2_log_severity& severity) const;
 
+        tf2_ros::Buffer _tf_buffer;
+        tf2_ros::TransformListener* _tf_listener;
```
Comment (ChristofDubs) [inline realsense2_camera/include/realsense_node_factory.h]: Is it necessary that this is a pointer, or can you just have an object as a member such that `new` allocation and calling `delete` is not required?

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 2 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: This is link to the formula: https://gamedev.stackexchange.com/questions/28395/rotating-vector3-by-a-quaternion

Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Gravity vector is such that because in the coordinate system assumed by Intel RS is X->pointing to right, Y->pointing up.  Z->Coming out of the plane. Hence gravity (pointing downwards) shows up as 0.0, -9.81, 0.0 m/s2. We are inverse transforming this coordinate system with the relation has with base_link.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 1 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp:25]: Done. Will commit the code.

File: realsense2_camera/include/realsense_node_factory.h (hunk 4 above)
Comment (sumandeepb) [reply realsense2_camera/include/realsense_node_factory.h]: Pointer is necessary because it needs to be initialized with the _tf_buffer object. As class member data, it cannot be initialized otherwise. I will see if we can move the listener to a member function local variable, then we can avoid pointer as well as allocation/deallocation.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 3 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp:191]: Done. Will commit the code.

File: realsense2_camera/include/realsense_node_factory.h (hunk 4 above)
Comment (sumandeepb) [reply realsense2_camera/include/realsense_node_factory.h]: Checked this: Both _tf_buffer and _tf_listener objects need to be static or class members. Making them static fails the code in the first attempt as it takes some time to initialize. So the sensor init code succeeds only in the second attempt or later. Making them class member and creating the object in the class constructor has the objects ready before the sensor initialization code. This seems to be the best option.

Comment (ChristofDubs) [reply realsense2_camera/include/realsense_node_factory.h]: Hmm in some other PR, I used a local `tf2_ros::TransformListener` and that seemed to work

Comment (gauthamm) [reply realsense2_camera/include/realsense_node_factory.h]: >  The TransformListener object should be scoped to persist otherwise it's cache will be unable to fill and almost every query will fail. A common method is to make the TransformListener object a member variable of a class.

[From ros wiki](http://wiki.ros.org/tf/Tutorials/Writing%20a%20tf%20listener%20%28C%2B%2B%29) . 
I remember having similar issue and needed to make the listener object persist rather than instantaneous creation/destruction

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 5]
```diff
@@ -30,6 +36,8 @@ RealSenseNodeFactory::RealSenseNodeFactory()
 		ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug);
 
 	rs2::log_to_console(severity);
+
+	_tf_listener = new tf2_ros::TransformListener(_tf_buffer);
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: would suggest using std::unique_ptr instead of handling raw pointers

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 6]
```diff
@@ -134,7 +254,12 @@ void RealSenseNodeFactory::onInit()
 #endif
 		ros::NodeHandle nh = getNodeHandle();
 		auto privateNh = getPrivateNodeHandle();
+
+		/* change: sumandeepb: determine correct serial no for realsense_top and realsense_bottom */
+		privateNh.param("camera", _camera_name, std::string(""));
+		privateNh.param("base_frame_id", _tf_link_name, std::string(""));
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: should this be base_frame_id or accel_frame_id ?

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 7]
```diff
@@ -57,22 +67,132 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
 		else
 		{
 			bool found = false;
+
+			if (_accel_orientation.empty()) // generate accel orientation vector from camera quaternion
+			{
+				try {
+					geometry_msgs::TransformStamped transformStamped = 
+									_tf_buffer.lookupTransform(_tf_link_name, "base_link", ros::Time(0));
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: I would suggest to get base_link also as a param called reference_frame. Though its not likely in short term, we might change the reference frame at some point to something else and its easier to catch it in launch file.

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 8]
```diff
@@ -57,22 +67,132 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
 				ROS_DEBUG_STREAM("Device with serial number " << sn << " was found.");
-				if (_serial_no.empty() || sn == _serial_no)
+
+				if (!_serial_no.empty() && sn == _serial_no)
 				{
 					_device = dev;
 					_serial_no = sn;
 					found = true;
 					break;
+				} else if (_serial_no.empty() && !_accel_orientation.empty()) // auto determine correct realsense orientation
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: It looks like the first `if` and `else` execute same set of instructions. Might be better to do `if (_serial_no.empty() && !_accel_orientation.empty()) { ... } else {...}`

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 9]
```diff
@@ -57,22 +67,132 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+						{
+							pipe.start(cfg);
+							success = true;
+							break;
+						} catch (...)
+						{
+							no_attempts ++;
+							ROS_DEBUG_STREAM("Pipe failed to create on device serial no " << sn << ". Device busy. Waiting..." << no_attempts);
+							ros::Duration(2).sleep();
+							continue;
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: you don't need a continue

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 10]
```diff
@@ -57,22 +67,132 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+						}
+					}
+
+					if (!success)
+					{
+						// all max attempts exhausted for the current device
+						ROS_INFO_STREAM("Pipe failed to create on device serial no " << sn << " after " << no_attempts << " attempts");
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp:197]: ROS_WARN_STREAM

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 11]
```diff
@@ -57,22 +67,132 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+
+					// attempt to read accel data frame for upto max attempts
+					no_attempts = 0;
+					while (no_attempts < MAX_ATTEMPTS)
+					{
+						rs2::frameset frameset = pipe.wait_for_frames();
+						no_attempts ++;
+
+						// fetch accelerometer orientation data from pipe stream
+						if (rs2::motion_frame accel_frame = frameset.first_or_default(RS2_STREAM_ACCEL))
```
Comment (gauthamm) [inline realsense2_camera/src/realsense_node_factory.cpp]: I assume you meant to set the value and compare null. Just verifying.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 6 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: accel_frame_id in TF tree is only available after the sensor is initialized. At this stage of execution only base_frame_id is available. Fortunately this has the same rotation as accel_frame_id just a slightly different translation. Therefore, determining orientation works out.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 5 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Alright. Will make the changes.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 7 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Alright. Will make the changes.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 8 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Alright. Will make the changes. I also thought this was becoming very cumbursome to understand.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 10 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp:197]: This is not anything serious. Just expected to happen in one RS ros node when the sensor is allocated correctly by other RS ros node. Serious is when both nodes fail to grab a sensor each, which is reported as ERROR at a later code line.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 11 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Yes. Maybe I will split the statements for better clarity.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 10 above)
Comment (gauthamm) [reply realsense2_camera/src/realsense_node_factory.cpp:197]: ok

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 6 above)
Comment (gauthamm) [reply realsense2_camera/src/realsense_node_factory.cpp]: hmm, can you add a todo/Note to the code here to explain the reasoning. Would be helpful to future readers to understand the reason.

**Comment by sumandeepb** [comment]
will add design details and coding decisions in this document: https://docs.google.com/document/d/1Dg6gszNfD36r9-z7hvgF-JKsA93_GhdvBopIPmeXlRo/

**Review by gauthamm — APPROVED** [review]

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 2 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. changed the code to be more clear and explainable. Added more comments to clear logic.

Still using quaternion based rotation transformation instead of 3X3 rot vec. Reason, getting some error in tf2::convert to 3X3 rot vec. It always returns in Q form. 

This code is one time execution during infrastructure boot. So overhead has no effect on robot running.

File: realsense2_camera/include/realsense_node_factory.h (hunk 4 above)
Comment (sumandeepb) [reply realsense2_camera/include/realsense_node_factory.h]: Added more clear comments to explain the design

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 5 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. Code committed.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 6 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Added more clear comments to explain the design

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 7 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. Code committed.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 8 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. Code committed.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 9 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. Code committed.

File: realsense2_camera/src/realsense_node_factory.cpp (hunk 11 above)
Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp]: Done. Code committed.

**Comment by sumandeepb** [comment]
@ChristofDubs made all requested changes. Kindly check and accept PR if alright.

File: realsense2_camera/src/realsense_node_factory.cpp [hunk 12]
```diff
@@ -57,12 +65,118 @@ void RealSenseNodeFactory::getDevice(rs2::device_list list)
...
+					// universal gravity vector in refence frame of robot in quaternion form
+					q_gravity_base.setValue(0.0, 0.0, 9.81);
+					// fetch pure rotation quaternion Q of sensor relative to robot reference frame
+					tf2::convert(transformStamped.transform.rotation , q_camera_rot);
+					// Calculate the new orientation of gravity in sensor base reference frame
+					// apply quaternion rotation to gravity vector q' = Q * q * Q^-1
+					// https://gamedev.stackexchange.com/questions/28395/rotating-vector3-by-a-quaternion
+					q_gravity_camera = q_camera_rot * q_gravity_base * q_camera_rot.inverse();
+					// transform to accel optical frame reference (uses different convention than robot reference base and camera base)
+					q_gravity_camera.setValue(-q_gravity_camera.y(), -q_gravity_camera.z(), q_gravity_camera.x());
```
Comment (ChristofDubs) [inline realsense2_camera/src/realsense_node_factory.cpp:87]: It actually says [here](https://github.com/IntelRealSense/librealsense/blob/master/doc/d435i.md) that the IMU frame is aligned with the depth (optical) frame, so you ideally you would lookup the transform from `base_link` to `camera_depth_optical_frame` to not have to do this. (You might also consider to add a 4th timeout argument to the lookupTransform call)

Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp:87]: "camera_depth_optical_frame" is also unavilable unless the sensor is initialized. Using the "base_frame_id" is the only available solution in the sensor initialization code.

[ERROR 1575618742.529330843 ros.realsense2_camera]: "tf_realsense_bottom_depth_optical_frame" passed to lookupTransform argument target_frame does not exist. 
[ERROR 1575618742.529602993 ros.realsense2_camera]: "tf_realsense_top_depth_optical_frame" passed to lookupTransform argument target_frame does not exist.

Comment (sumandeepb) [reply realsense2_camera/src/realsense_node_factory.cpp:87]: Any suggestions for the timeout duration ? How about 2 secs ?

Comment (ChristofDubs) [reply realsense2_camera/src/realsense_node_factory.cpp:87]: Hmm it seems like that urdf file isn't actually used, apart from some test launch files?

Ok i just checked the code and there is literally a hard-coded `quaternion_optical.setRPY(-M_PI / 2, 0.0, -M_PI / 2);` statement in `calcAndPublishStaticTransform()` ... So I guess we can lower the bar a bit on not hard-coding things.

**Review by ChristofDubs — APPROVED** [review]

**Review by sumandeepb — COMMENTED** [review]
Branch working fine on actual robot without issues.

Also tested on laptop with different combinations of connected and missing RealSense plus tried out orientation based and serial no based device assignment.
